public/blog/birthday02/article.md

```diff
@@ -0,0 +1,462 @@
+
+This caused plenty of pain-points around type checking. For example, the rough flow of checking, if a `for` loop is type-safe and valid, is the following.
+
+ 1. Check what the type of the iterated collection is, call it `TCollection`
+ 2. Check, if `TCollection` has a method called `GetEnumerator()`, which returns a type `TEnumerator`
+ 3. Check, if `TEnumerator` has a method called `MoveNext()`, which returns a `bool`
+ 4. Check, if `TEnumerator` has a property called `Current`, which returns a type `TElement`
+ 5. Check, if `Telement` is assignable to the type of the loop variable, in case it's explicitly typed
+
+Imagine, that we could not infer the exact type of the collection yet. How do we check if it has a method called `GetEnumerator()`? We can't! The solution? Introducing various kinds of sentinel, delay and placeholder nodes in the untyped tree, with the sole purpose of pushing back the check, evaluation or node construction to the point, where we have more information available from future code. With this mentality, the untyped tree basically became an incomplete copy of the bound tree, with node such as `UntypedLocalExpression` or `UntypedDelayedExpression`.
+
+This caused multiple maintainability issues. First off, binding had to be written essentially twice. Once when translating the syntax tree to untyped trees, and then the untyped trees to bound trees - with the constraint solver invoked in between for more type information. Second, the code became multiply nested with callbacks. It was not uncommon to see this:
+
+```cs
+var sequenceExpr = new UntypedDelayedExpression(sequence.Type, () =>
+{
+    var memberConstraint = constraintSolver.MemberConstraint(sequence.Type, "GetEnumerator");
+    return new UntypedDelayedExpression(memberConstraint, () =>
+    {
+        // ...
+    });
+});
+```
+
+And finally, if we ever had to do type-checks on a node, we had to take into account that it could be one of the placeholders or delays anytime, and had to wrap that logic itself into a delay node to deal with this.
+
+If you have ever seen [asynchronous JavaScript code](https://www.stoman.me/articles/async-await-promises-callbacks-in-javascript) before async/await, you might get a very familiar feeling. This is the exact same problem! When we don't know something yet, it would be awesome to suspend the evaluation of the current node, continue binding other nodes, or even evaluate some constraints in the constraint solver. I told this to [Kuinox](https://github.com/Kuinox/), who immediately got to work, and not even an hour later presented the prototype: binding tasks, that work exactly like JavaScript async callbacks.
```

Comment:
"you might get a very familiar feeling" is an odd way to say that in English. But I am not sure how to rephrase it.

Reply:
Replaced with "this might look familiar to you", Idk if it's an improvement